### ISSUE TITLE
On the recent i386 overflow errors

### ISSUE THREAD

**Issue by LemonBoy** [ziglang/zig#8924]
As you've probably noticed the CI is failing with the following output:
```
thread 27376 panic: integer overflow
/home/vsts/work/1/s/lib/std/mem.zig:2690:41: 0x678393 in mem.alignForwardGeneric (test)
    return alignBackwardGeneric(T, addr + (alignment - 1), alignment);
                                        ^
/home/vsts/work/1/s/lib/std/heap.zig:1157:42: 0x6ef539 in heap.testAllocatorAlignedShrink (test)
    slice = try allocator.reallocAdvanced(slice, mem.page_size * 32, alloc_size / 2, .exact);
                                         ^
/home/vsts/work/1/s/lib/std/heap.zig:980:35: 0x542e4d in heap.test "std-i386-linux-gnu-Debug-c-multi StackFallbackAllocator" (test)
    try testAllocatorAlignedShrink(stack_allocator.get());
                                  ^
/home/vsts/work/1/s/lib/std/special/test_runner.zig:76:28: 0x66d997 in std.special.main (test)
        } else test_fn.func();
                           ^
/home/vsts/work/1/s/lib/std/start.zig:458:37: 0x69956f in start.callMain (test)
            const result = root.main() catch |err| 
                                    ^
/home/vsts/work/1/s/lib/std/start.zig:400:12: 0x66f9dc in start.callMainWithArgs (test)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/home/vsts/work/1/s/lib/std/start.zig:370:12: 0x66f6a9 in start.main (test)
    return @call(.{ .modifier = .always_inline }, callMainWithArgs, .{ @intCast(usize, c_argc), c_argv, envp });
           ^
```
When running i386 binaries (or, I guess, ARM32 binaries) on 64bit systems we may have to handle pointers in the upper-half region of the address space and, in that case, be prepared to handle overflows and underflows when evaluating the base pointer alignment in `heap.zig` (and possibly other places in the stdlib).

A few problems need to be solved:
- [ ] The alignment functions must handle the overflow/underflow case and signal it by returning an error
- [ ] All the call-sites must be updated
- [ ] The `heap.zig` logic must be updated to avoid, detect and report overflows/underflows

**Comment by andrewrk** [comment]
I see, so it's testing a large alignment, which potentially cannot be fulfilled if the pointer is too close to the end of the address space.

**Comment by LemonBoy** [comment]
Yes. Dropping the `i386-linux-gnu` target certainly fixes the problem.

**Comment by andrewrk** [comment]
Are you suggesting to keep the tests enabled even though they have intermittent failures?

I think our policy of "no flaky tests, no exceptions" is reasonable. As long as the failure is represented by an open issue, correctly labeled "bug", and correctly blocking a release that would make it a regression, we're all good.

For this issue, if we get it done in time for 0.8.0 that's great, but it does not block the release because it's not a regression from any of the prior releases.

**Comment by marnix** [comment]
The 0.8.0 release notes say (about duplicate #8930): "This issue is not solved in 0.8.0 but is planned to be addressed in the 0.8.1 bug fix release."

So the milestone for this issue should be updated, it looks like?

**Comment by andrewrk** [comment]
>  Also fix #8924 to some degree, the amount of unchecked alignForward is still scary.

Alrighty then, renaming this issue and re-opening
